unicode in parameter names

yurii-github opened the pull request:
for https://github.com/yiisoft/yii2/issues/8508
just to show what i mean

---

<b>similar</b>
#4446 (will fail if source is not in utf8)

---

<b>notes</b>
- htmlspecialchars hit a limit of available charsets for code files
- code becomes very heavy on check if other charset than utf-8  is supported
- gives ability to use url with non ascii chars like  index.php?єжыъэ=эъхж , in fact, it's a must in some cases when non ascii attribute is translated into query string via GET

---

<b>current</b>

```
// Yii::$app->charset = 'cp1251', project files are encoded in cp1251 accordingly     
$f = ActiveForm::begin();
echo $f->field($ar, 'рік');
ActiveForm::end();
```

output looks like

```
<form id="w0" action=".." method="post"><div class="form-group field-test-рік">
<label class="control-label" for="test-рік">Рік</label>
<input type="text" id="test-рік" class="form-control" name="Test[рік]" value="2010">

<div class="help-block"></div>
</div></form>
```

yurii-github committed:
unicode in parameter names

yurii-github committed:
fix for utf-8

yurii-github committed:
fix utf-8

samdark commented on framework/helpers/BaseHtml.php:
```diff
@@ -2040,7 +2040,7 @@ public static function getInputName($model, $attribute)
     public static function getInputId($model, $attribute)
     {
-        $name = strtolower(static::getInputName($model, $attribute));
+        $name = mb_strtolower(static::getInputName($model, $attribute));
```
Encoding should be passed when using `mb_`  i.e. `mb_strtolower($string, Yii::$app->charset);`.

samdark commented on framework/helpers/BaseInflector.php:
```diff
@@ -317,7 +317,7 @@ public static function camelize($word)
     public static function camel2words($name, $ucwords = true)
     {
-        $label = trim(strtolower(str_replace([
+        $label = trim(mb_strtolower(str_replace([
```
Same here.

yurii-github replied:
yes. you are right. but tests have not failed

yurii-github committed:
added encoding

yurii-github committed:
added encoding

yurii-github committed:
oops

yurii-github committed:
added mb_convert, simplified label

samdark commented on framework/helpers/BaseInflector.php:
```diff
@@ -317,13 +317,12 @@ public static function camelize($word)
-        ], ' ', preg_replace('/(?<![A-Z])[A-Z]/', ' \0', $name))));
-
-        return $ucwords ? ucwords($label) : $label;
+        $label = preg_replace('/(?<![A-Z])[A-Z]/', ' \0', $name); // add spaces TODO: unicode?
```
Yes, this one is not unicode-aware...

yurii-github committed:
unit test change

with unicode adds dependency on app object which is not good

yurii-github committed:
added other encoding support

adds bottle neck like transformation from user encoding to utf-8 for preg_match and back

yurii-github committed:
oops

yurii-github commented:
i managed to get it going, but i'm not sure it worth the candles.
such approach requires dependency on app object which is not good for static method.

yurii-github commented:
i looked a bit at code more. does this 'Yii::charset" do anything at all except sending useless contentType header?

imagine example: my code base is in cp1251, yii2 in utf8. i set Yii::app->charset = 'cp1251', db connection to cp1251 too. now yii2 - dies or lives?

i tried it recently on cp1251 code files. in config i put charset to cp1251.

     echo yii\helpers\Html::encode('фівїхфжівї');
     echo '<br>';
     echo 'фівїхфжівї';

output was with right cp1251 encoding, but Html encode failed

     пїЅпїЅпїЅпїЅпїЅ
     фівїхфжівї

how this charset option is supposed to work?

regards

yurii-github committed:
_GET to non utf8 code

translates non ascii chars from utf8 url to code encoding

example
index.php?фів=їзхщєдж

yurii-github commented:
it has huge chain.. it looks like better to abandon this

samdark commented:
Hmm... indeed it looks too complicated. Let's leave this pull request open for a while. Need to think about it.

Commit:
enc fix getAttributeName, getAttributeValue, getInputName
normalize conversion

Commit:
protected -> public

yurii-github commented:
alright. whenever you please. i'll be pushing some from time to time with a study purpose.

Commit:
getQueryParams() enc

yurii-github commented:
i have a question about code -  $this->_queryParams is never reached in Request:: getQueryParams() because this var is never set. also  this class is 0% covered with tests.

what's the purpose of method Request::setQueryParams() ?

klimov-paul commented:
`Request::setQueryParams()` can be used in unit tests.

cebe commented on framework/helpers/BaseHtml.php:
```diff
@@ -89,6 +89,35 @@ class BaseHtml
+        }
+        return mb_convert_encoding($encString, 'UTF-8', Yii::$app->charset);
+    }
+    
```
why do we need these encoding conversions? shouldn't the whole applicaion shere the same encoding?

yurii-github replied:
if file encode is different from utf8
1 - preg_match with /u expects unicode
2 - $_GET parses data from url as unicode because url is in utf8

unfortunately, i don't know how to preg_match non unicode and unicode without conversion

cebe replied:
every file in the project should be utf8 encoded. if not we would go crazy try to catch all the problems this causes.

yurii-github replied:
hm
what's the point in Yii::app->charset? it must be forced to utf8 then. or i don't get something

cebe replied:
if you change the charset, you should change it overall. So there still should be no need to convert anything.

yurii-github replied:
i don't understand what "if you change the charset, you should change it overall" means

here's sample code which will fail (and that's right): 
yii charset in config = windows-1251 (Cyrillic)  / 'charset' => 'windows-1251'
code files are encoded in utf-8
sample layout outputs meta as <meta charset="<?= Yii::$app->charset ?>">
we want output 
echo 'year : рік'; 

when you open in IE you will see 
year : СЂС–Рє 

regards

cebe replied:
why would you set application charset to windows-1251 and have the files in utf-8?

yurii-github replied:
to show you that setting yii charset other than utf-8 fails unless we convert project files into set encoding (windows-1251 etc). and doing so will fail my modification above if i remove string conversions to utf-8 and vice versa.

yurii-github commented:
klimov-paul,
why do so? PHPUnit resets super globals on each test

yurii-github commented:
this pull doesnt make any sense so far. closing

regards
